core/src/SYCL/Kokkos_SYCL_Parallel_Scan.hpp

```diff
@@ -227,23 +220,30 @@ class ParallelScanSYCLBase {
   void impl_execute(const PostFunctor& post_functor) {
     if (m_policy.begin() == m_policy.end()) return;
 
-    const auto& q = *m_policy.space().impl_internal_space_instance()->m_queue;
+    auto& instance        = *m_policy.space().impl_internal_space_instance();
     const std::size_t len = m_policy.end() - m_policy.begin();
 
-    // FIXME_SYCL The allocation should be handled by the execution space
-    // consider only storing one value per block and recreate initial results in
-    // the end before doing the final pass
-    auto deleter = [&q](value_type* ptr) { sycl::free(ptr, q); };
-    std::unique_ptr<value_type[], decltype(deleter)> result_memory(
-        static_cast<pointer_type>(sycl::malloc(sizeof(value_type) * len, q,
-                                               sycl::usm::alloc::shared)),
-        deleter);
-    m_scratch_space = result_memory.get();
+    // compute the total amount of memory we will need.
+    std::size_t total_memory = 0;
+    {
+      size_t wgroup_size   = 32;
+      size_t n_nested_size = len;
+      size_t n_nested_wgroups;
+      do {
+        total_memory += sizeof(value_type) * n_nested_size;
+        n_nested_wgroups = (n_nested_size + wgroup_size - 1) / wgroup_size;
+        n_nested_size    = n_nested_wgroups;
+      } while (n_nested_wgroups > 1);
+      total_memory += sizeof(value_type) * wgroup_size;
+    }
```

Comment:
Comment on what is going on here.

Reply:
[`df8843a` (#3899)](https://github.com/kokkos/kokkos/pull/3899/commits/df8843a15f0924c395879d8bf8621f0cf54c7624)